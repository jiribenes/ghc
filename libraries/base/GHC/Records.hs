{-# LANGUAGE AllowAmbiguousTypes #-}
{-# LANGUAGE FunctionalDependencies #-}
{-# LANGUAGE PolyKinds #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TypeApplications #-}

-----------------------------------------------------------------------------
-- |
-- Module      :  GHC.Records
-- Copyright   :  (c) Adam Gundry 2015-2016
-- License     :  see libraries/base/LICENSE
--
-- Maintainer  :  cvs-ghc@haskell.org
-- Stability   :  internal
-- Portability :  non-portable (GHC extensions)
--
-- This module defines the 'HasField' class used by the
-- @OverloadedRecordFields@ extension.  See the
-- <https://gitlab.haskell.org/ghc/ghc/wikis/records/overloaded-record-fields
-- wiki page> for more details.
--
-----------------------------------------------------------------------------

module GHC.Records
       ( HasField(..)
       , getField
       , setField
       ) where

-- | Constraint representing the fact that the field @x@ can be get and set on
--   the record type @r@ and has field type @a@.  This constraint will be solved
--   automatically, but manual instances may be provided as well.
--
--   The function should satisfy the invariant:
--
-- > uncurry ($) (hasField @x r) == r

--   HasField :: forall {k}. k -> * -> * -> Constraint
--   getField :: forall {k} (x::k) r a. HasField x r a => r -> a
-- NB: The {k} means that k is an 'inferred' type variable, and
--     hence not provided in visible type applications.  Thus you
--     say     getField @"foo"
--     not     getField @Symbol @"foo"
class HasField x r a | x r -> a where
  -- | Function to get and set a field in a record.
  hasField :: r -> (a -> r, a)

-- | Selector function to extract the field from the record.
getField :: forall x r a . HasField x r a => r -> a
getField = snd . hasField @x

-- | Update function to set a new value of the field on the record.
setField :: forall x r a . HasField x r a => r -> a -> r
setField = fst . hasField @x
