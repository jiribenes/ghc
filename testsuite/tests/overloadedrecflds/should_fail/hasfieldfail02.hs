{-# LANGUAGE DataKinds, ExistentialQuantification, MagicHash, RankNTypes,
             TypeApplications #-}

import GHC.Prim (Int#)
import GHC.Records (HasField, getField)

data T = MkT { foo :: forall a . a -> a }
data U = forall b . MkU { bar :: b }
data V = MkV { baz :: Int# }

-- This should fail because foo is higher-rank.
x = getField @"foo" (MkT id)

-- This should fail because bar is a naughty record selector (it
-- involves an existential).
y = getField @"bar" (MkU True)

-- This should fail because baz is not of kind Type.
z = getField @"baz" (MkV 3#)

main = return ()
