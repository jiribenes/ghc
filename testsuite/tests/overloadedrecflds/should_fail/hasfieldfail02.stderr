
hasfieldfail02.hs:12:5: error:
    • No instance for (HasField "foo" T a2)
        arising from a use of ‘getField’
    • In the expression: getField @"foo" (MkT id)
      In an equation for ‘x’: x = getField @"foo" (MkT id)

hasfieldfail02.hs:16:5: error:
    • No instance for (HasField "bar" U a1)
        arising from a use of ‘getField’
    • In the expression: getField @"bar" (MkU True)
      In an equation for ‘y’: y = getField @"bar" (MkU True)

hasfieldfail02.hs:19:5: error:
    • No instance for (HasField "baz" V a0)
        arising from a use of ‘getField’
    • In the expression: getField @"baz" (MkV 3#)
      In an equation for ‘z’: z = getField @"baz" (MkV 3#)
